instrumentation/opentelemetry_req/lib/opentelemetry_req.ex

```diff
@@ -0,0 +1,206 @@
+defmodule OpentelemetryReq do
+  @moduledoc """
+  Wraps the request in an opentelemetry span. Span names must be paramaterized, so the
+  `req_path_params` module and step should be registered before this step. This step is
+  expected by default and an error will be raised if the path params option is
+  not set for the request.
+
+  Given the steps pipeline can be halted to skip further steps from running, it is important
+  to _append_ request and response steps _after_ this step to ensure execution. Spans are not
+  created until the request is completed or errored.
+
+  ### Example
+
+  ```
+  client =
+    Req.new()
+    |> OpentelemetryReq.attach()
+    |> Req.Request.merge_options(
+      base_url: "http://localhost:4000",
+      propagate_trace_ctx: true
+    )
+
+  client
+  |> Req.get(
+    url: "/api/users/:user_id",
+    path_params: [user_id: user_id]
+  )
+  ```
+  ## Request Options
+
+    * `:span_name` - `String.t()` if provided, overrides the span name. Defaults to `nil`.
+    * `:no_path_params` - `boolean()` when set to `true` no path params are expected for the request. Defaults to `false`
+    * `:propagate_trace_ctx` - `boolean()` when set to `true`, trace headers will be propagated. Defaults to `false`
+  """
+
+  alias OpenTelemetry.Tracer
+  alias OpenTelemetry.SemanticConventions.Trace
+  require Trace
+  require Tracer
+  require Logger
+
+  def attach(%Req.Request{} = request, options \\ []) do
+    request
+    |> Req.Request.register_options([:span_name, :no_path_params, :propagate_trace_ctx])
+    |> Req.Request.merge_options(options)
+    |> Req.Request.append_request_steps(
+      require_path_params: &require_path_params_option/1,
+      start_span: &start_span/1,
+      put_trace_headers: &maybe_put_trace_headers/1
+    )
+    |> Req.Request.prepend_response_steps(otel_end_span: &end_span/1)
+    |> Req.Request.prepend_error_steps(otel_end_span: &end_errored_span/1)
+  end
+
+  defp start_span(request) do
+    span_name = span_name(request)
+
+    attrs = build_req_attrs(request)
+
+    parent_ctx = OpenTelemetry.Ctx.get_current()
+
+    Tracer.start_span(span_name, %{
+      attributes: attrs,
+      kind: :client
+    })
+    |> Tracer.set_current_span()
+
+    Req.Request.put_private(request, :otel_parent_ctx, parent_ctx)
```

Comment:
I don't see this used. Did you mean to use it to set the context back to `parent_ctx`?

Reply:
Nah. Forgot to delete that. Seems to work fine without juggling.

Reply:
Eh, it should return to the previous context when it finishes, since it is starting a new one and setting it to the current in the pdict it'd be issue back in the code that called Req.

Reply:
So end span would return the previous context when done from `Tracer` and then needs to be attached again?

Reply:
The spans currently seem to behave as expected now.

Reply:
Ok. Think this is all correct now.